### ISSUE TITLE
AudioSource Doppler effect creates artefacts

### ISSUE THREAD

**Issue by giovannibedetti** [rorywalsh/CsoundUnity#35]
Workaround for now is setting the Doppler Level to 0, but that disables the effect completely.

**Comment by sinoth** [comment]
Is this a limitation in doing procedural audio in general? Is the idea here that the doppler plugin doesn't have enough look ahead information from the audio source to meaningfully pitch something up?

**Comment by giovannibedetti** [comment]
Yes to your first question, no to the second: it's a general limitation of procedural audio, but lookahead isn't what's missing.

As far as I know, Unity's doppler works by changing playback rate. On a recorded clip that genuinely is a doppler — reading stored samples faster raises the pitch. On a procedural source it can't be: Csound generates in real time, at the output rate, exactly the block Unity asks for, so varying the rate just time-warps audio that already exists rather than making the oscillator run higher. The rate tracks relative velocity and keeps changing, so you get continuous unstable repitching that follows the movement but sounds nothing like a physical doppler. There's no stored buffer to look ahead into, by the time the rate change applies, the sound has already been produced.

The fix belongs in Csound: send velocity, azimuth, elevation and rolloff as channels and let the CSD do the work — hrtfmove2 for the binaural part, and the doppler opcode, which takes source and mic positions and reads a delay line at a delay computed from their distance. Physically correct, and it's already the pattern CsoundUnity uses on WebGL where Unity's spatialization isn't available at all with generated audio.

One thing worth trying: the upcoming 4.0 adds new audio output paths, and IAudioGenerator (Unity 6+) drives the AudioSource directly and skips Unity's resampling step — so this may simply not happen there. I haven't verified it. If you want to test, it's on the [release_4_0_0](https://github.com/rorywalsh/CsoundUnity/tree/release_4_0_0) branch — still work in progress, but usable. Let me know how it behaves if you do.

Until then: dopplerLevel = 0.

**Comment by sinoth** [comment]
Thanks for the explanation! I'm specifically interested in positional audio in VR, which has the added complexity of spatialization done via the Meta SDK. I'm not super clear where in the signal path this resides, though I think it sits at the end and still uses Unity doppler if enabled.

**Comment by giovannibedetti** [comment]
Went and actually tested this before giving you a proper answer, and it turned out to be worth doing (disclaimer: tested on editor only for now)

Setup: Meta XR Audio SDK 78 on Unity 6, CsoundUnity driving a 7.1.4 speaker rig: one Csound instance writing 12 named channels, one CsoundUnityChild per speaker.

Good news first: the combination works nicely. Meta's spatializer sits at the end of the chain and receives Csound's output, so HRTF, room acoustics and distance attenuation all apply to procedurally generated audio just like any other source. Two things have to be right, and both are easy to miss: spatializePostEffects must be true, which CsoundUnity sets for you, and AudioSource.spatialize must be on, without it the Meta plugin never enters the chain at all and you just get plain Unity panning.

Doppler is the part that still needs work. It does reach the output, so my earlier guess that Unity 6's IAudioGenerator path sidesteps the resampling was wrong, ignore that one. Whether what you get is a correct doppler is what I'm still measuring: at realistic speeds the shift is only a percent or two, and "this sounds wrong" and "this sounds right but subtler than I expected" are genuinely hard to tell apart by ear. I'd rather come back with numbers than impressions.

Either way the Csound-side approach still looks like the better answer, a modelled doppler beats a resampled one.
